content/lectures/lec02.md

```diff
@@ -7,3 +7,190 @@ draft: false
+VERSION_CODENAME=focal
+UBUNTU_CODENAME=focal
+```
+
+We can see that the Docker container is running a fresh install of Ubuntu. Now, we could do all manner of things with this Ubuntu container: install new software, expose ports, connect to a network, run a web server, etc. However, actually doing these things manually within the Ubuntu terminal is _not_ reproducible; we would have to perform those same operations every time we wanted to spin up a new container. Hence, a Dockerfile. When we write a Dockerfile it is essentially a roadmap for how the Docker image should be run, so that all of the installation and configuration of whatever toolchain we need on our container can be written out concisely once and only once.
+
+However, feel free to try out different commands or installing different software on the Ubuntu container. It works the same as any other Linxu command line interface and there is a lot you can do with it. You can type the `exit` comand to leave the Ubuntu terminal at any time.
+
+### Building an App on Docker
+
+Let's build a simple Node web application on Docker to outline the development process. First, create a new directory anywhere on your machine called `node_demo`. Now within this directory will write all of our application code. First we'll start with a Dockerfile that has all of the requisite steps for setting up our Docker application:
+
+```
+# Pull from latest verision of Node
+FROM node:latest
+
+# Set the working directory to this directory (common place to 
+# store application in the Linux file structure)
+WORKDIR /usr/src/app
+
+# Copy over package.json and packaage-lock.json (these files
+# contain the necessary dependencies for our project)
+COPY package*.json ./
+
+# Install dependencies for our project
+RUN npm install
+
+# Copy over all of the app source code
+COPY . .
+
+# Expose the port on which we will run our application
+EXPOSE 8080
+
+# Start the server
+CMD ["node", "index.js"]
+```
+
+Before we carry on to the next step, there is a little nuance here that we should underscore. Note that we copy over the `packagae.json` and `package-lock.json` files before installing our dependencies. Why don't we just install the dependencies and then copy everything over at once? The answer is that Docker is smart and if the files `packagae.json` and `package-lock.json` have not changed, it can skip the following build steps where it installs the dependencies and copies them over. This works because the two files `packagae.json` and `package-lock.json` contain all of the Node dependencies, so if they have not changed we know that the dependencies have not changed.
+
+Now that we've written our Dockerfile, we actually should initialize a Node application (if you don't have Node installed, you can find an download guide on their [website](https://nodejs.org/en/download/)). Once you have Node installed, we can initialize the project within our `node_demo` (that also containers our Dockerfile):
+
+```
```

Comment:
this is a lot of stuff to have them write down. can we just have a directory in `website` with all the files already there so they can just focus on actually running the docker commands and changing things line-by-line?